Make ddm_quota calculation quadratic and normalized

haozturk opened the pull request:
This PR

1. Makes ddm_quota quadratic to dynamic + free fraction instead of linear. This is because we want RSEs with low dynamic + free fraction chosen with less probability compared to the current implementation
2. Normalizes the weights between 0 and 100
3. Makes the script a bit more modular

One thing to note
 - I do linear (Max-Min) normalization. This means that the RSE with the lowest dynamic + free fraction will have 0 ddm_quota

And one question: Can someone explain or point me the relevant code exactly how Rucio uses this weight?

Cheers

dynamic-entropy commented on docker/CMSRucioClient/scripts/updateDDMQuota:
```diff
@@ -10,43 +10,76 @@ client = Client()
+
+    calculate_ddm_quotas()
+    normalize_ddm_quotas()
+    if not DRY_RUN:
```
The purpose of dry_runsis to print out the changes without actually doing it.
You can pass the variable and check it within the set_ddm_quotas function and only set it when DRY_RUN is false.

haozturk replied:
Indeed, thanks for catching it. I fixed it.

dynamic-entropy commented on docker/CMSRucioClient/scripts/updateDDMQuota:
```diff
@@ -10,43 +10,76 @@ client = Client()
+def normalize_ddm_quotas():
+    weights = [int(value) for value in ddm_quotas.values()]
+    for rse, weight in ddm_quotas.items():
+        ddm_quotas[rse] = int( ((int(weight) - min(weights)) / (max(weights) - min(weights)) ) * MAX_DDM_QUOTA )
```
This normalisation would always put ddm_quota for RSE with min weight to 0. 
Is this intentional?

haozturk replied:
Not intentional. It's just a feature of this normalization. I don't think setting the lowest weight to 0 would have any negative operational consequences.

dynamic-entropy commented on docker/CMSRucioClient/scripts/updateDDMQuota:
```diff
@@ -10,43 +10,76 @@ client = Client()
+        # Apprantly, python integers do not overflow, https://docs.python.org/3/library/exceptions.html#OverflowError
+
+        locked = rucio - unavailable - expired
+        ddm_quota = int((max(static - locked, 0) / static) * 1e4) ** 2
```
Given this is being squared. The multiplication with 10k is not required.

haozturk replied:
We should multiple it by a number >= 10, o/w integer casting converts all values to 0, since they are between 0 and 1.

dynamic-entropy reviewed (changes requested):
Now that we are at it. I believe we should try to find a way to give higher weight to RSEs with empty space as compared to those with dynamic space. 
What do you think?

dynamic-entropy commented:
Additionally, can you do a dry_run and paste its output?

haozturk commented:
> Now that we are at it. I believe we should try to find a way to give higher weight to RSEs with empty space as compared to those with dynamic space. What do you think?

Just to understand: Let's say there are two RSEs whose Rucio quotas are the same where 

For RSE1; free space is 2X TB and dynamic space is X is TB
For RSE2; free space is X and the dynamic space is 2X is TB

Are you proposing to give higher weight to RSE1 compared to RSE2 since it has bigger free space? If so, it makes sense to me. We can put a higher coefficient to the free space in the calculation such as

```
ddm_quota = N * free_space + dynamic space

where N > 1
```

What do you think?

haozturk commented:
I've made changes in the dry-run logic and here is the output after:
```
$ python3 docker/CMSRucioClient/scripts/updateDDMQuota 
DRY-RUN: Set ddm_quota for T2_US_Purdue to 13
DRY-RUN: Set ddm_quota for T2_FI_HIP to 53
DRY-RUN: Set ddm_quota for T2_FR_GRIF_IRFU to 95
DRY-RUN: Set ddm_quota for T2_PL_Swierk to 15
DRY-RUN: Set ddm_quota for T1_FR_CCIN2P3_Disk to 34
DRY-RUN: Set ddm_quota for T2_UA_KIPT to 36
DRY-RUN: Set ddm_quota for T2_CH_CSCS to 32
DRY-RUN: Set ddm_quota for T2_US_UCSD to 0
DRY-RUN: Set ddm_quota for T2_BE_UCL to 22
DRY-RUN: Set ddm_quota for T2_IT_Rome to 26
DRY-RUN: Set ddm_quota for T2_IN_TIFR to 75
DRY-RUN: Set ddm_quota for T2_KR_KISTI to 45
DRY-RUN: Set ddm_quota for T2_US_Nebraska to 10
DRY-RUN: Set ddm_quota for T2_AT_Vienna to 9
DRY-RUN: Set ddm_quota for T2_PT_NCG_Lisbon to 17
DRY-RUN: Set ddm_quota for T2_RU_INR to 6
DRY-RUN: Set ddm_quota for T1_US_FNAL_Disk to 4
DRY-RUN: Set ddm_quota for T2_UK_SGrid_RALPP to 15
DRY-RUN: Set ddm_quota for T2_BE_IIHE to 41
DRY-RUN: Set ddm_quota for T2_EE_Estonia to 13
DRY-RUN: Set ddm_quota for T2_PK_NCP to 100
DRY-RUN: Set ddm_quota for T2_BR_SPRACE to 29
DRY-RUN: Set ddm_quota for T2_FR_GRIF_LLR to 98
DRY-RUN: Set ddm_quota for T2_RU_ITEP to 12
DRY-RUN: Set ddm_quota for T2_UK_SGrid_Bristol to 20
DRY-RUN: Set ddm_quota for T2_PL_Cyfronet to 3
DRY-RUN: Set ddm_quota for T2_US_Caltech to 12
DRY-RUN: Set ddm_quota for T2_ES_IFCA to 44
DRY-RUN: Set ddm_quota for T2_CH_CERN to 0
DRY-RUN: Set ddm_quota for T1_IT_CNAF_Disk to 27
DRY-RUN: Set ddm_quota for T2_DE_RWTH to 38
DRY-RUN: Set ddm_quota for T1_RU_JINR_Disk to 28
DRY-RUN: Set ddm_quota for T2_DE_DESY to 23
DRY-RUN: Set ddm_quota for T2_CN_Beijing to 24
DRY-RUN: Set ddm_quota for T1_UK_RAL_Disk to 44
DRY-RUN: Set ddm_quota for T2_RU_IHEP to 4
DRY-RUN: Set ddm_quota for T2_ES_CIEMAT to 40
DRY-RUN: Set ddm_quota for T2_UK_London_Brunel to 42
DRY-RUN: Set ddm_quota for T1_ES_PIC_Disk to 29
DRY-RUN: Set ddm_quota for T2_US_Vanderbilt to 5
DRY-RUN: Set ddm_quota for T2_US_MIT to 4
DRY-RUN: Set ddm_quota for T1_DE_KIT_Disk to 20
DRY-RUN: Set ddm_quota for T2_TW_NCHC to 42
DRY-RUN: Set ddm_quota for T2_TR_METU to 30
DRY-RUN: Set ddm_quota for T2_UK_London_IC to 15
DRY-RUN: Set ddm_quota for T2_US_Wisconsin to 1
DRY-RUN: Set ddm_quota for T2_IT_Bari to 1
DRY-RUN: Set ddm_quota for T2_IT_Legnaro to 19
DRY-RUN: Set ddm_quota for T2_IT_Pisa to 37
DRY-RUN: Set ddm_quota for T2_FR_IPHC to 58
DRY-RUN: Set ddm_quota for T2_HU_Budapest to 36
DRY-RUN: Set ddm_quota for T2_RU_JINR to 42
DRY-RUN: Override ddm_quota for T2_BR_UERJ to 0
DRY-RUN: Override ddm_quota for T2_US_Florida to 1
DRY-RUN: Override ddm_quota for T2_FR_GRIF to 0
```

dynamic-entropy commented:
> What do you think?
Yes, sounds good. Please add that too.
N=2 should be good enough.

haozturk commented:
I added that, plus got rid of code duplication, introduced better error handling and increased the range to 1000. See the dry run output below.

I still have doubts about this. The fact that the whole calculation is relative to the total volume of the RSE might not be optimal. We might need to consider introducing another factor in the equation to boost ddm_quota for larger sites. This depends how dynamically Rucio can adapt to the changes in the weight. Currently Nebraska has 1.5PB of free+dynamic space whereas FNAL has more than 10PB while Nebraska's weight is much higher due to its relatively smaller size. Let's say we have a rule of 1PB data with ddm_quota as weight which started sending data to Nebraska. During the transfer, will Rucio be able to understand that Nebraska's ddm_quota is getting smaller, thus it should switch to FNAL for the rest of the transfer? Perhaps an update as follows might make sense:

```
 ddm_quota =  (static / (<sum of all statics>)) * (((FREE_SPACE_COEFFICIENT * free + expired) / static) * 10) ** 2
```
Let me know what you think
```
$ python3 docker/CMSRucioClient/scripts/updateDDMQuota
Skipping T2_US_MIT_Buffer due to lack of relevant key in rse
{'unavailable', 'static', 'expired', 'rucio'} is not a subset of dict_keys(['expired', 'obsolete', 'rucio', 'unavailable'])
DRY-RUN: Set ddm_quota for T2_ES_IFCA to 133
DRY-RUN: Set ddm_quota for T2_US_Caltech to 93
DRY-RUN: Set ddm_quota for T1_IT_CNAF_Disk to 79
DRY-RUN: Set ddm_quota for T2_BE_UCL to 69
DRY-RUN: Set ddm_quota for T2_KR_KISTI to 141
DRY-RUN: Set ddm_quota for T2_IT_Rome to 120
DRY-RUN: Set ddm_quota for T2_US_MIT to 4
DRY-RUN: Set ddm_quota for T2_UK_SGrid_RALPP to 50
DRY-RUN: Set ddm_quota for T2_US_Vanderbilt to 18
DRY-RUN: Set ddm_quota for T2_HU_Budapest to 109
DRY-RUN: Set ddm_quota for T2_PK_NCP to 597
DRY-RUN: Set ddm_quota for T2_CH_CSCS to 99
DRY-RUN: Set ddm_quota for T2_PL_Swierk to 49
DRY-RUN: Set ddm_quota for T2_UK_SGrid_Bristol to 51
DRY-RUN: Set ddm_quota for T2_US_Purdue to 46
DRY-RUN: Set ddm_quota for T2_US_Nebraska to 34
DRY-RUN: Set ddm_quota for T2_TW_NCHC to 126
DRY-RUN: Set ddm_quota for T2_PT_NCG_Lisbon to 88
DRY-RUN: Set ddm_quota for T2_DE_DESY to 93
DRY-RUN: Set ddm_quota for T2_AT_Vienna to 1
DRY-RUN: Set ddm_quota for T2_IT_Pisa to 116
DRY-RUN: Set ddm_quota for T2_ES_CIEMAT to 124
DRY-RUN: Set ddm_quota for T1_UK_RAL_Disk to 133
DRY-RUN: Set ddm_quota for T2_BR_SPRACE to 116
DRY-RUN: Set ddm_quota for T2_FR_GRIF_IRFU to 972
DRY-RUN: Set ddm_quota for T2_UK_London_Brunel to 89
DRY-RUN: Set ddm_quota for T2_CH_CERN to 4
DRY-RUN: Set ddm_quota for T2_TR_METU to 94
DRY-RUN: Set ddm_quota for T2_IN_TIFR to 223
DRY-RUN: Set ddm_quota for T2_US_Wisconsin to 8
DRY-RUN: Set ddm_quota for T1_DE_KIT_Disk to 73
DRY-RUN: Set ddm_quota for T2_UA_KIPT to 113
DRY-RUN: Set ddm_quota for T2_RU_ITEP to 35
DRY-RUN: Set ddm_quota for T2_FI_HIP to 525
DRY-RUN: Set ddm_quota for T2_EE_Estonia to 61
DRY-RUN: Set ddm_quota for T1_ES_PIC_Disk to 91
DRY-RUN: Set ddm_quota for T1_RU_JINR_Disk to 88
DRY-RUN: Set ddm_quota for T2_PL_Cyfronet to 10
DRY-RUN: Set ddm_quota for T2_UK_London_IC to 66
DRY-RUN: Set ddm_quota for T2_US_UCSD to 4
DRY-RUN: Set ddm_quota for T2_IT_Legnaro to 86
DRY-RUN: Set ddm_quota for T2_RU_JINR to 131
DRY-RUN: Set ddm_quota for T2_RU_IHEP to 41
DRY-RUN: Set ddm_quota for T2_BE_IIHE to 123
DRY-RUN: Set ddm_quota for T2_CN_Beijing to 77
DRY-RUN: Set ddm_quota for T1_FR_CCIN2P3_Disk to 109
DRY-RUN: Set ddm_quota for T2_FR_IPHC to 173
DRY-RUN: Set ddm_quota for T2_DE_RWTH to 118
DRY-RUN: Set ddm_quota for T2_IT_Bari to 16
DRY-RUN: Set ddm_quota for T2_RU_INR to 25
DRY-RUN: Set ddm_quota for T2_FR_GRIF_LLR to 1000
DRY-RUN: Set ddm_quota for T1_US_FNAL_Disk to 6
DRY-RUN: Override ddm_quota for T2_BR_UERJ to 0
DRY-RUN: Override ddm_quota for T2_FR_GRIF to 0
DRY-RUN: Override ddm_quota for T2_US_Florida to 1
```

dynamic-entropy commented:
Thankyou @haozturk 
Why is the ddm_quota for FNAL and CERN so low in this?
I think we need to play more with the function to get something more desirable. 
We can try your above suggestion too?

dynamic-entropy commented:
Yeah, this looks better. 
Please turn off the verbose flag. I will merge it. 


(Now I wonder, I different this is from the previous numbers).

haozturk commented:
Thanks done

haozturk commented:
Btw, I don't like adding +1 to sites with `ddm_quota` with 0. It's 0, because it's the worst site to send data according to your equation. So, I see no problem with not sending data there. IMO, relative space usage plot should be fixed so that it demonstrates sites with ddm_quota 0 as well.
